File: content/snippets/cpp/design_patterns/object_pool.cpp
#include <vector>
#include <memory>
#include <iostream>

class PooledObject {
public:
    int value;
};

class ObjectPool {
    std::vector<std::unique_ptr<PooledObject>> pool;
public:
    PooledObject* acquire() {
        if (!pool.empty()) {
            auto p = pool.back().release(); pool.pop_back();
            return p;
        }
        return new PooledObject();
    }
    void release(PooledObject* obj) {
        pool.emplace_back(obj);
    }
};

int main(){
    ObjectPool p;
    auto *o = p.acquire();
    o->value = 42;
    std::cout<<o->value<<"\n";
    p.release(o);
}

File: content/snippets/go/design_patterns/object_pool.go
package main

import "fmt"

type PooledObject struct{ Value int }

type ObjectPool struct{ pool chan *PooledObject }

func NewObjectPool(size int) *ObjectPool {
	return &ObjectPool{pool: make(chan *PooledObject, size)}
}
func (p *ObjectPool) Acquire() *PooledObject {
	select {
	case o := <-p.pool:
		return o
	default:
		return &PooledObject{}
	}
}
func (p *ObjectPool) Release(o *PooledObject) {
	select {
	case p.pool <- o:
	default:
	}
}

func main() {
	p := NewObjectPool(2)
	o := p.Acquire()
	o.Value = 5
	fmt.Println(o.Value)
	p.Release(o)
}
